sklearn/tree/_criterion.pyx

```diff
@@ -599,6 +600,109 @@ cdef class Gini(ClassificationCriterion):
+
+    be the partial hellinger distance score for the k1 node.
+
+        weight_k2 = (\sqrt{\frac{N_{right_0}}{N_{parent_0}}}-\sqrt{\frac{N_{right_1}}{N_{parent_1}}})^2
+
+    be the partial hellinger distance score for the k2 node.
+
+    Hellinger distance score is then defined as:
+
+        hellinger_distance = \sqrt{weight_k1+weight_k2}
+
+    Note: hellinger distance range is [0, \sqrt{2}]
+
+    """
+    cdef double node_impurity(self) nogil:
+        """Evaluate the impurity of the current node.
+        
+        Evaluate the Hellinger distance criterion as impurity of the current node,
+        i.e. the impurity of samples[start:end]. The smaller the impurity the
+        better.
+        """
+        cdef:
+            double impurity_left = 0.0
+            double impurity_right = 0.0
+            double impurity_sum = 0.0
+
+        # get the partial hellinger score, separately per child
+        self.children_impurity(&impurity_left, &impurity_right)
+
+        # hellinger score is sqrt of children impurity sum
+        impurity_sum = impurity_right + impurity_left
+        if impurity_sum >= 0:
+            impurity_sum = sqrt(impurity_sum)
+
+        return impurity_sum
+
+    cdef void children_impurity(self, double* impurity_left,
+                                double* impurity_right) nogil:
+        """Evaluate the impurity in children nodes.
+        
+        i.e. the impurity of the left child (samples[start:pos]) and the
+        impurity the right child (samples[pos:end]) using Hellinger distance.
+
+        Parameters
+        ----------
+        impurity_left : double pointer
+            The memory address to save the impurity of the left node to
+        impurity_right : double pointer
+            The memory address to save the impurity of the right node to
+        """
+        cdef:
+            # taking only first sum_left and sum_right because only single label is supported
+            double[::1] sum_left = self.sum_left[0]
+            double[::1] sum_right = self.sum_right[0]
+
+            # sum of same class in both children nodes
+            double sum_k1 = sum_left[0] + sum_right[0]
+            double sum_k2 = sum_left[1] + sum_right[1]
+
+            # hellinger distance score calculation helpers
+            double weight_k1_left = 0.0
+            double weight_k2_left = 0.0
+            double weight_k1_right = 0.0
+            double weight_k2_right = 0.0
+            double hellinger_distance_left = 0.0
+            double hellinger_distance_right = 0.0
+
+        # in case any of the classes sum is 0 it means that
+        # the classes are perfectly separated and no additional split is required
+        if  sum_k1 > 0 and sum_k2 > 0:
+            weight_k1_left = sqrt(sum_left[0] / sum_k1)
+            weight_k1_right = sqrt(sum_right[0] / sum_k1)
+            weight_k2_left = sqrt(sum_left[1] / sum_k2)
+            weight_k2_right = sqrt(sum_right[1] / sum_k2)
```

Comment:
What if `sum_k1` or `sum_k2` is null?

Reply:
`sum_k1` is assigned here `double sum_k1 = sum_left[0] + sum_right[0]` thus it can't be null, it will crash before the assignment
`sum_left[0]` is assigned here `double[::1] sum_left = self.sum_left[0]` and I guess can potentially get a null in case `self.sum_left[0]` has null pointer inside
but it is a class member that is widely used, and not verified for null in other places e.g. in gini 

I totally agree that the code should be as safe as possible and all corner cases should be validated
In case we do it first in Hellinger, I'm not sure how the code should through an exception from Cython if the check is positive and the member is indeed null

please share your thoughts